src/Rings/mpoly-ideals.jl

```diff
@@ -347,6 +347,11 @@ Ideal generated by
 ```
 """
 function saturation(I::MPolyIdeal{T}, J::MPolyIdeal{T} = ideal(base_ring(I), gens(base_ring(I))); iteration::Bool=false) where T
+  # `remove` seems to be faster than `Singular.saturation`
+  if ((ngens(I) == 1) & (ngens(J) == 1))
```

Comment:
```suggestion
  if ngens(I) == 1 && ngens(J) == 1
```
Why not _logical or_ and get rid of lots of parentheses?

Reply:
[ederc](https://github.com/ederc) left the extra parentheses, and yes, these can be removed.

I use & instead of && because && is slower, it creates an extra branching point.

Reply:
What is the runtime different? Does `&` also support short-circuiting?

Reply:
There is not a single such use of ` & ` in the Oscar repository, so I don't think this can be considered standard.

My main issue with `&` is what you experienced with the first commit, julia has a different operator precedence for `&` than, e.g., C, which can lead to very subtle and difficult to spot errors.
I do care more about maintainability than speed for this line.

Since you do seem to care about benchmarking and micro-optimizations:
```julia
julia> fun(I,J) = if ngens(I) == 1 && ngens(J) == 1
       foo()
       else
       bar()
       end
fun (generic function with 1 method)

julia> fun2(I,J) = if (ngens(I) == 1) & (ngens(J) == 1)
       foo()
       else
       bar()
       end
fun2 (generic function with 1 method)

julia> foo() = 1; bar() = 0;

julia> R,(x,y,z) = polynomial_ring(QQ, [:x, :y, :z]);

julia> I = J = ideal(x);

julia> @benchmark fun(I,J)
BenchmarkTools.Trial: 10000 samples with 998 evaluations.
 Range (min … max):  13.653 ns … 36.132 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     14.965 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   15.393 ns ±  1.939 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▄▂▂▄▅▅█▅▅▆▄▃ ▂                                              ▁
  ███████████████▆▇▇▆▆▅▅▅▄█▅▅▆▄▅▅▅▂▃▅▆▅▅▅▄▅▄▄▄▄▂▃▃▄▄▄▅▅▇▇██▇▇ █
  13.7 ns      Histogram: log(frequency) by time        25 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark fun2(I,J)
BenchmarkTools.Trial: 10000 samples with 998 evaluations.
 Range (min … max):  13.448 ns … 35.484 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     15.292 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   15.487 ns ±  1.522 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

         ▁▂█▇▄▁                                                
  ▃▃▄▄▃▅▆██████▆▇▃▄▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂ ▃
  13.4 ns         Histogram: frequency by time        24.3 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

And for the case where `&&` can short-circuit:
```
julia> I = ideal([x,y])
Ideal generated by
  x
  y

julia> @benchmark fun(I,J)
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  12.382 ns … 29.325 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     14.118 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   14.083 ns ±  0.676 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▂  ▄▃  ▁▂▁  ▁▂▂  ▂▄▃   ▆▅▅ ▂█▇▄▂▅▃▆▁▅▃▂   ▃                 ▂
  █▁▁██▆▁████████▆▇██████████████████████▇▇▆██████▆▆▆█▆▇▆▇▆▄▆ █
  12.4 ns      Histogram: log(frequency) by time      15.9 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark fun2(I,J)
BenchmarkTools.Trial: 10000 samples with 998 evaluations.
 Range (min … max):  13.871 ns … 55.766 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     15.299 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   15.300 ns ±  0.990 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▂     ▄▃ ▃▁▁  ▅▃▅▁ ▃▃█▄   ▆▅          ▂                     ▂
  █▄▁▃▆▆██▁████▆████▇████▇▃▄██▆▆▇▇▅▅▆█▆▆█▆▆▇▆▇▆▇▆▅▅▅▄▆▆▅▆▆▄▄▆ █
  13.9 ns      Histogram: log(frequency) by time      17.8 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
(with dummy functions instead of `remove` / `saturation` which would otherwise dominate the timings)
So maybe there is a very tiny advantage for the short-circuit, but these timings do vary by a few ns. Even the fastest invocation of `remove` that I could come up with using two different variables is already 10 times slower than this. And `saturation` is way slower.

Note that even for such a very simple ideal in a rational polynomial ring `ngens` does look involved (with quite a few branches and calls):
```julia
julia> @code_native debuginfo=:none ngens(I)
	.text
	.file	"number_of_generators"
	.globl	julia_number_of_generators_194319 # -- Begin function julia_number_of_generators_194319
	.p2align	4, 0x90
	.type	julia_number_of_generators_194319,@function
julia_number_of_generators_194319:      # @julia_number_of_generators_194319
; Function Signature: number_of_generators(Oscar.MPolyIdeal{Nemo.QQMPolyRingElem})
# %bb.0:                                # %top
	push	rbp
	mov	rbp, rsp
	push	rbx
	sub	rsp, 24
	vxorps	xmm0, xmm0, xmm0
	vmovaps	xmmword ptr [rbp - 32], xmm0
	mov	qword ptr [rbp - 16], 0
	#APP
	mov	rax, qword ptr fs:[0]
	#NO_APP
	mov	rbx, qword ptr [rax - 8]
	mov	qword ptr [rbp - 32], 4
	mov	rax, qword ptr [rbx]
	mov	qword ptr [rbp - 24], rax
	lea	rax, [rbp - 32]
	mov	qword ptr [rbx], rax
	#DEBUG_VALUE: number_of_generators:I <- [DW_OP_deref] undef
	#DEBUG_VALUE: number_of_generators:I <- [DW_OP_deref] undef
	mov	rax, qword ptr [rdi]
	test	rax, rax
	je	.LBB0_9
# %bb.1:                                # %pass
	mov	rcx, qword ptr [rax]
	test	rcx, rcx
	je	.LBB0_10
# %bb.2:                                # %pass4
	mov	rax, qword ptr [rcx + 32]
	test	rax, rax
	je	.LBB0_7
# %bb.3:                                # %pass8
	movabs	rcx, offset ".Ljl_global#194327.jit"
	mov	rdx, qword ptr [rcx]
	mov	rcx, qword ptr [rdx + 5736]
	test	rcx, rcx
	je	.LBB0_11
# %bb.4:                                # %pass40
	mov	rsi, qword ptr [rax]
	mov	rdi, qword ptr [rdx + 5744]
	mov	qword ptr [rbp - 16], rax
	call	rcx
	cdqe
	jmp	.LBB0_5
.LBB0_7:                                # %L57
	mov	rax, qword ptr [rcx + 8]
	test	rax, rax
	je	.LBB0_6
# %bb.8:                                # %pass50
	mov	rax, qword ptr [rax + 16]
.LBB0_5:                                # %L64
	mov	rcx, qword ptr [rbp - 24]
	mov	qword ptr [rbx], rcx
	add	rsp, 24
	pop	rbx
	pop	rbp
	ret
.LBB0_9:                                # %fail
	movabs	rax, offset jl_undefref_exception
	mov	rdi, qword ptr [rax]
	movabs	rax, offset ijl_throw
	call	rax
.LBB0_10:                               # %fail3
	movabs	rax, offset jl_undefref_exception
	mov	rdi, qword ptr [rax]
	movabs	rax, offset ijl_throw
	call	rax
.LBB0_11:                               # %fail39
	movabs	rcx, offset jl_undefref_exception
	mov	rdi, qword ptr [rcx]
	mov	qword ptr [rbp - 16], rax
	movabs	rax, offset ijl_throw
	call	rax
.LBB0_6:                                # %fail49
	movabs	rax, offset jl_undefref_exception
	mov	rdi, qword ptr [rax]
	movabs	rax, offset ijl_throw
	call	rax
.Lfunc_end0:
	.size	julia_number_of_generators_194319, .Lfunc_end0-julia_number_of_generators_194319
                                        # -- End function
.set ".Ljl_global#194327.jit", 139774091562048
	.size	".Ljl_global#194327.jit", 8
	.section	".note.GNU-stack","",@progbits
```

Reply:
Difference between & and && is mostly subjective, up to personal preference. If OSCAR style guide recommends using && instead of &, please give a link to the relevant point. Otherwise, thank you for your suggestion, but I prefer the way I wrote it

Reply:
According to
https://docs.julialang.org/en/v1/manual/control-flow/#Short-Circuit-Evaluation
`&` does not support short circuiting. Hence `&&` is better for most logical operations -- at least in our context (I am sure a counterexample where `&` is faster can be cooked up).
Upshot: please use `&&` for short circuit and for uniformity of the code.

Reply:
I think "better" here is very subjective. If you really think that && should be used everywhere in code for logical operations, please (bring up an issue to) add this to the style guide. Until then, thank you for your suggestion, but I prefer the way I wrote it

Reply:
We can discuss this in person next week.